From stern@rowland.harvard.edu Wed Aug 22 10:06:58 2007
From: Alan Stern <stern@rowland.harvard.edu>
Date: Wed, 22 Aug 2007 13:06:53 -0400 (EDT)
Subject: USB: fix mistake in usb_hcd_giveback_urb
To: Greg KH <greg@kroah.com>
Message-ID: <Pine.LNX.4.44L0.0708221305450.8838-100000@iolanthe.rowland.org>


This patch (as971) fixes a small mistake: The URB's completion status
needs to be adjusted before the URB is passed to usmon_urb_complete(),
not afterward.

Signed-off-by: Alan Stern <stern@rowland.harvard.edu>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 drivers/usb/core/hcd.c |    7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

--- a/drivers/usb/core/hcd.c
+++ b/drivers/usb/core/hcd.c
@@ -1266,9 +1266,6 @@ int usb_hcd_unlink_urb (struct urb *urb,
  */
 void usb_hcd_giveback_urb (struct usb_hcd *hcd, struct urb *urb)
 {
-	unmap_urb_for_dma(hcd, urb);
-	usbmon_urb_complete (&hcd->self, urb);
-	usb_unanchor_urb(urb);
 	urb->hcpriv = NULL;
 	if (unlikely(urb->unlinked))
 		urb->status = urb->unlinked;
@@ -1277,6 +1274,10 @@ void usb_hcd_giveback_urb (struct usb_hc
 			!urb->status))
 		urb->status = -EREMOTEIO;
 
+	unmap_urb_for_dma(hcd, urb);
+	usbmon_urb_complete(&hcd->self, urb);
+	usb_unanchor_urb(urb);
+
 	/* pass ownership to the completion handler */
 	urb->complete (urb);
 	atomic_dec (&urb->use_count);
